### PR TITLE
Add `list_channels_by_counterparty` method

### PR DESCRIPTION
**Pull request by tnull**
While we already provide a `list_channels` method, it could result in quite a large `Vec<ChannelDetails>`. Here, we provide the means to query our channels by `counterparty_node_id` and DRY up the code.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -1229,6 +1229,55 @@ impl ChannelDetails {
 	pub fn get_outbound_payment_scid(&self) -> Option<u64> {
 		self.short_channel_id.or(self.outbound_scid_alias)
 	}
+
+	fn from_channel<Signer: ChannelSigner + WriteableEcdsaChannelSigner>(channel: &Channel<Signer>,
```
Comment (arik-so) [inline lightning/src/ln/channelmanager.rs:1233]: FYI, WriteableEcdsaChannelSigner inherits ChannelSigner. That said, absolutely love this method

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -1786,6 +1794,25 @@ where
 		self.list_channels_with_filter(|&(_, ref channel)| channel.is_live())
 	}
 
+	/// Gets the list of channels we have with a given counterparty, in random order.
+	pub fn list_channels_with_counterparty(&self, counterparty_node_id: &PublicKey) -> Vec<ChannelDetails> {
```
Comment (arik-so) [inline lightning/src/ln/channelmanager.rs:1798]: is this significantly faster than just doing a .filter on list_channels?

File: lightning/src/ln/channelmanager.rs [hunk 3]
```diff
@@ -1786,6 +1794,25 @@ where
...
+		let mut res = Vec::with_capacity(self.short_to_chan_info.read().unwrap().len());
+		{
```
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:1800]: Scope not needed.

File: lightning/src/ln/channelmanager.rs [hunk 4]
```diff
@@ -1786,6 +1794,25 @@ where
 		self.list_channels_with_filter(|&(_, ref channel)| channel.is_live())
 	}
 
+	/// Gets the list of channels we have with a given counterparty, in random order.
+	pub fn list_channels_with_counterparty(&self, counterparty_node_id: &PublicKey) -> Vec<ChannelDetails> {
+		let mut res = Vec::with_capacity(self.short_to_chan_info.read().unwrap().len());
```
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:1799]: Won't this allocate too much? You should use `peer_state.channel_by_id.len()` instead.

Also, I *think* if you use `iter().map(...).collect()`, it's smart enough to allocated enough capacity.

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 6562d30a..042ff9d2 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -1796,21 +1796,19 @@ where
 
        /// Gets the list of channels we have with a given counterparty, in random order.
        pub fn list_channels_with_counterparty(&self, counterparty_node_id: &PublicKey) -> Vec<ChannelDetails> {
-               let mut res = Vec::with_capacity(self.short_to_chan_info.read().unwrap().len());
-               {
-                       let best_block_height = self.best_block.read().unwrap().height();
-                       let per_peer_state = self.per_peer_state.read().unwrap();
-                       if let Some(peer_state_mutex) = per_peer_state.get(counterparty_node_id) {
-                               let mut peer_state_lock = peer_state_mutex.lock().unwrap();
-                               let peer_state = &mut *peer_state_lock;
-                               for (_channel_id, channel) in peer_state.channel_by_id.iter() {
-                                       let details = ChannelDetails::from_channel(channel, best_block_height,
-                                               peer_state.latest_features.clone());
-                                       res.push(details);
-                               }
-                       }
-               }
-               res
+               let best_block_height = self.best_block.read().unwrap().height();
+               let per_peer_state = self.per_peer_state.read().unwrap();
+               if let Some(peer_state_mutex) = per_peer_state.get(counterparty_node_id) {
+                       let mut peer_state_lock = peer_state_mutex.lock().unwrap();
+                       let peer_state = &mut *peer_state_lock;
+                       let features = &peer_state.latest_features;
+                       return peer_state.channel_by_id
+                               .iter()
+                               .map(|(_, channel)|
+                                       ChannelDetails::from_channel(channel, best_block_height, features.clone()))
+                               .collect();
+               }
+               vec![]
        }
 
        /// Returns in an undefined order recent payments that -- if not fulfilled -- have yet to find a
```

**Review by jkczyz — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**